File: packages/raisin-client-java/src/main/java/com/raisindb/client/utils/RequestTracker.java
package com.raisindb.client.utils;

import com.raisindb.client.exceptions.RequestException;
import com.raisindb.client.exceptions.TimeoutException;
import com.raisindb.client.protocol.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Map;
import java.util.UUID;
import java.util.concurrent.*;
import java.util.function.Consumer;

/**
 * Tracks pending requests and resolves them when responses arrive.
 */
public class RequestTracker {

    private static final Logger logger = LoggerFactory.getLogger(RequestTracker.class);

    private final Consumer<byte[]> sendFunc;
    private final long defaultTimeout;
    private final Map<String, CompletableFuture<Object>> pendingRequests;
    private final ScheduledExecutorService scheduler;
    private final MessagePackCodec codec;

    public RequestTracker(Consumer<byte[]> sendFunc, long defaultTimeout) {
        this.sendFunc = sendFunc;
        this.defaultTimeout = defaultTimeout;
        this.pendingRequests = new ConcurrentHashMap<>();
        this.scheduler = Executors.newScheduledThreadPool(2);
        this.codec = new MessagePackCodec();
    }

    /**
     * Send a request and wait for the response.
     *
     * @param requestType Type of request
     * @param context     Request context
     * @param payload     Request payload
     * @return CompletableFuture with the response result
     */
    public CompletableFuture<Object> sendRequest(
            RequestType requestType,
            RequestContext context,
            Object payload
    ) {
        return sendRequest(requestType, context, payload, defaultTimeout);
    }

    /**
     * Send a request with custom timeout.
     */
    public CompletableFuture<Object> sendRequest(
            RequestType requestType,
            RequestContext context,
            Object payload,
            long timeout
    ) {
        String requestId = UUID.randomUUID().toString();
        CompletableFuture<Object> future = new CompletableFuture<>();

        // Store pending request
        pendingRequests.put(requestId, future);

        // Set timeout
        ScheduledFuture<?> timeoutHandle = scheduler.schedule(() -> {
            CompletableFuture<Object> pending = pendingRequests.remove(requestId);
            if (pending != null && !pending.isDone()) {
                pending.completeExceptionally(
                        new TimeoutException("Request " + requestId + " timed out"));
            }
        }, timeout, TimeUnit.MILLISECONDS);

        // Cancel timeout when future completes
        future.whenComplete((result, error) -> timeoutHandle.cancel(false));

        try {
            // Create and send request envelope
            RequestEnvelope envelope = new RequestEnvelope(requestId, requestType, context, payload);
            byte[] data = codec.encodeRequest(envelope);
            sendFunc.accept(data);

        } catch (Exception e) {
            pendingRequests.remove(requestId);
            future.completeExceptionally(e);
        }

        return future;
    }

    /**
     * Handle an incoming response.
     */
    public void handleResponse(ResponseEnvelope response) {
        CompletableFuture<Object> future = pendingRequests.remove(response.getRequestId());

        if (future == null) {
            // Response for unknown request (might have timed out)
            logger.warn("Received response for unknown request: {}", response.getRequestId());
            return;
        }

        if (response.getStatus() == ResponseStatus.ERROR) {
            ErrorInfo error = response.getError();
            if (error != null) {
                future.completeExceptionally(
                        new RequestException(error.getMessage(), error.getCode(), error.getDetails()));
            } else {
                future.completeExceptionally(
                        new RequestException("Unknown error", "UNKNOWN"));
            }
        } else if (response.getStatus() == ResponseStatus.SUCCESS ||
                   response.getStatus() == ResponseStatus.COMPLETE) {
            future.complete(response.getResult());
        } else if (response.getStatus() == ResponseStatus.STREAMING) {
            // For now, just return the first chunk
            // TODO: Implement proper streaming support
            future.complete(response.getResult());
        }
    }

    /**
     * Shutdown the request tracker.
     */
    public void shutdown() {
        scheduler.shutdown();
        try {
            if (!scheduler.awaitTermination(5, TimeUnit.SECONDS)) {
                scheduler.shutdownNow();
            }
        } catch (InterruptedException e) {
            scheduler.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }

    public MessagePackCodec getCodec() {
        return codec;
    }
}

File: packages/raisin-client-py/raisindb/utils/request_tracker.py
"""Request tracking and promise resolution."""

import asyncio
from typing import Any, Callable, Dict, Optional
from uuid import uuid4

from ..protocol import (
    RequestEnvelope,
    ResponseEnvelope,
    ResponseStatus,
    RequestType,
    RequestContext,
)
from ..exceptions import RequestError, TimeoutError


class PendingRequest:
    """Represents a pending request awaiting response."""

    def __init__(self, request_id: str, timeout: float):
        """Initialize a pending request."""
        self.request_id = request_id
        self.future: asyncio.Future[Any] = asyncio.Future()
        self.timeout = timeout
        self.timeout_handle: Optional[asyncio.TimerHandle] = None

    def set_result(self, result: Any) -> None:
        """Set the successful result."""
        if self.timeout_handle:
            self.timeout_handle.cancel()
        if not self.future.done():
            self.future.set_result(result)

    def set_error(self, error: Exception) -> None:
        """Set an error result."""
        if self.timeout_handle:
            self.timeout_handle.cancel()
        if not self.future.done():
            self.future.set_exception(error)

    def set_timeout(self) -> None:
        """Mark the request as timed out."""
        if not self.future.done():
            self.future.set_exception(TimeoutError(f"Request {self.request_id} timed out"))


class RequestTracker:
    """Tracks pending requests and resolves them when responses arrive."""

    def __init__(
        self,
        send_func: Callable[[bytes], None],
        default_timeout: float = 30.0,
    ):
        """
        Initialize the request tracker.

        Args:
            send_func: Function to send encoded request bytes
            default_timeout: Default timeout for requests in seconds
        """
        self.send_func = send_func
        self.default_timeout = default_timeout
        self.pending_requests: Dict[str, PendingRequest] = {}
        self._lock = asyncio.Lock()

    async def send_request(
        self,
        request_type: RequestType,
        context: RequestContext,
        payload: Any,
        timeout: Optional[float] = None,
    ) -> Any:
        """
        Send a request and wait for the response.

        Args:
            request_type: Type of request
            context: Request context
            payload: Request payload
            timeout: Timeout in seconds (uses default if None)

        Returns:
            The response result

        Raises:
            RequestError: If the request fails
            TimeoutError: If the request times out
        """
        request_id = str(uuid4())
        timeout = timeout or self.default_timeout

        # Create pending request
        pending = PendingRequest(request_id, timeout)

        async with self._lock:
            self.pending_requests[request_id] = pending

        # Set timeout
        loop = asyncio.get_event_loop()
        pending.timeout_handle = loop.call_later(
            timeout, lambda: self._handle_timeout(request_id)
        )

        try:
            # Create and send request envelope
            from ..protocol import encode_request

            envelope = RequestEnvelope(
                request_id=request_id,
                request_type=request_type,
                context=context,
                payload=payload,
            )

            self.send_func(encode_request(envelope))

            # Wait for response
            result = await pending.future
            return result

        finally:
            # Clean up
            async with self._lock:
                self.pending_requests.pop(request_id, None)

    async def handle_response(self, response: ResponseEnvelope) -> None:
        """
        Handle an incoming response.

        Args:
            response: The response envelope
        """
        async with self._lock:
            pending = self.pending_requests.get(response.request_id)

        if not pending:
            # Response for unknown request (might have timed out)
            return

        if response.status == ResponseStatus.ERROR:
            if response.error:
                error = RequestError(
                    response.error.message, response.error.code, response.error.details
                )
                pending.set_error(error)
            else:
                pending.set_error(RequestError("Unknown error", "UNKNOWN"))
        elif response.status in (ResponseStatus.SUCCESS, ResponseStatus.COMPLETE):
            pending.set_result(response.result)
        elif response.status == ResponseStatus.STREAMING:
            # For now, we'll just return the first chunk
            # TODO: Implement proper streaming support
            pending.set_result(response.result)

    def _handle_timeout(self, request_id: str) -> None:
        """Handle request timeout."""

        async def _do_timeout() -> None:
            async with self._lock:
                pending = self.pending_requests.get(request_id)
                if pending:
                    pending.set_timeout()

        asyncio.create_task(_do_timeout())
